### PR TITLE
Speed up arg_to_iter

### LINKED ISSUE #50: The re-introduction of nested item support caused a significant performance degradation
I have a CPU-bound Scrapy project that becomes 50% slower after https://github.com/scrapy/itemloaders/pull/29.

I believe the problem is that `is_item` is not a cheap call, and it will potentially become more expensive as itemadapter extends support to additional types.

I think we may be able to reimplement this function so that it does not call `is_item` at all, but still does not treat item-like objects as sequences.

### PR DESCRIPTION
**Pull request by Gallaecio**
Fixes #50

If we restrict the subset of sequence and generator types that we allow `arg_to_iter` to handle, we can remove the use of `is_item` altogether to significantly improve its performance.

I tested performance with:

```python
from timeit import timeit

from itemloaders.utils import arg_to_iter


def a():
    return arg_to_iter({})

print(timeit(a, number=1000000))
```

Results:
- Before https://github.com/scrapy/itemloaders/pull/29, which introduced the issue: 0.1361095430002024
- Before this patch: 41.40795021699887
- After this patch: 0.5537120330009202

### CODE REVIEW

**Comment by Gallaecio** [comment]
The documentation issue is caused by https://github.com/scrapy/scrapy/issues/5402

**Comment by wRAR** [comment]
So I guess this changes the result from true to false for some types not listed in isinstance?

**Comment by Gallaecio** [comment]
It does.

I think supporting only a limited number of iterables as input for values is a low price for this much of a performance improvement. To be honest, I would have even dropped generator support be it not for the existing tests for it.

I think it would be best to document the specific iterables that are supported as input for values, limit the implementation to that subset, and let users do the required conversion for scenarios where custom types are needed.

Not a strong opinion, though. But since `is_item` is more likely to support a wider range of values, and likely to become slower as well as a result, I think avoiding its usage here is key for performance.

**Review by wRAR — APPROVED** [review]

File: itemloaders/utils.py [hunk 1]
```diff
@@ -19,14 +15,9 @@ def arg_to_iter(arg):
     """
     if arg is None:
         return []
-    elif (
-        hasattr(arg, '__iter__')
-        and not isinstance(arg, _ITERABLE_SINGLE_VALUES)
-        and not is_item(arg)
```
Comment (kmike) [inline itemloaders/utils.py:25]: It seems the docstring is no longer correct; it was also not really correct before the change as well. Would it make sense to fix it before merging?

**Comment by ava7** [comment]
> It does.
> 
> I think supporting only a limited number of iterables as input for values is a low price for this much of a performance improvement. To be honest, I would have even dropped generator support be it not for the existing tests for it.
> 
> I think it would be best to document the specific iterables that are supported as input for values, limit the implementation to that subset, and let users do the required conversion for scenarios where custom types are needed.
> 
> Not a strong opinion, though. But since `is_item` is more likely to support a wider range of values, and likely to become slower as well as a result, I think avoiding its usage here is key for performance.

@Gallaecio Any chance of keeping the `set` primitive? :pleading_face:

**Comment by Gallaecio** [comment]
Seems reasonable to me, feel free to open a PR.

**Pull request merged**